Starting to draft primal HHO

amartinhuertas opened the pull request:
WIP ...

amartinhuertas committed:
Starting to draft primal HHO

santiagobadia commented on draft_hho.jl:
```diff
@@ -0,0 +1,143 @@
+a(u,v)=∫(∇(v)⋅∇(u))dΩ
+l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ
+
+op=LocalAffineFEOperator((a,l),UKR,VKR)
```
```julia
# Define HHO local reconstruction operator
m(ur,v) = ∫∇(ur)*∇(v)dΩ
l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ
P=LocalProjector((m,l),UKR,VKR)

# Global formulation
UK∂K = MultiFieldFESpace([UK, U∂K])
a(u,v)=∫(∇(P(v))⋅∇(P(u)))dΩ

function f(v)
   vK, v∂K = UK∂K
   f(v) = ∫(f*vK)dΩ
end

# as HDG
```

santiagobadia commented on draft_hho.jl:
```diff
@@ -0,0 +1,143 @@
+
+a(u,v)=∫(∇(v)⋅∇(u))dΩ
+l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ
+
```
```julia
# Stabilisation
# Method 1: Orthogonal basis using change of basis with M^{-1}

refferecᵤ = ReferenceFE(orthogonal,Float64,order+1;space=:P)
\mu_T = h^{-2} # k^4
m(u,v)=∫(∇(v)⋅∇(u))dΩ + \mu_T ∫(v⋅u)dΩ
l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ + \mu_T ∫(v⋅u∂K)dΩ

# Optimisation
# \pi(k,u) that makes 0 dofs higher than k

# Method 2: Define subspaces of orthogonal spaces
# orthogonal basis from 1 to order+1 (excluding 0)
refferecᵤ = ReferenceFE(orthogonal,Float64,1:order+1;space=:P)
refferecᵤ_0 = ReferenceFE(orthogonal,Float64,0;space=:P)
VKR_1 = ...
VKR_0 = ...

VKR = MultiFieldFESpace([UKR_0,UKR_1])

# Define HHO local reconstruction operator
function m(u,v )
  v0, v1 = v
  m(u,v) = ∫∇(u)*∇(v1)dΩ +  ∫(u*v0) dΩ
end

function l(u,v )
  v0, v1 = v
  l( (uK,u∂K), v)=∫((∇(v1)⋅nK)*u∂K)d∂K+∫(-Δ(v1)*uK)dΩ + ∫(uK*v0) dΩ
end

P=LocalProjector((m,l),UKR,VKR)


m(ur,v) = ∫u_mean*vdΩ
l(uK, v)=∫v*uKdΩ
\pi_0=LocalProjector((m,l),UKR,VKR)



# Global formulation
UK∂K = MultiFieldFESpace([UK, U∂K])
a(u,v)=∫(∇(P(v))⋅∇(P(u)))dΩ

function f(v)
   vK, v∂K = UK∂K
   f(v) = ∫(f*vK)dΩ
end
# 


m(u,v)=∫(∇(v)⋅∇(u))dΩ 
l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ 

∫(v⋅u∂K)dΩ
```

santiagobadia commented on draft_hho.jl:
```diff
@@ -0,0 +1,143 @@
+l( (uK,u∂K), v)=∫((∇(v)⋅nK)*u∂K)d∂K+∫(-Δ(v)*uK)dΩ
+
+op=LocalAffineFEOperator((a,l),UKR,VKR)
+UKR_basis=op(get_trial_fe_basis.((UK,U∂K))...)
```
```julia
# Stab in Badia et al 2.11
# Define spaces for the projectors

# Cell term
m(ur,v) = ∫u*vdΩ
l( (uK,u∂K), v)=∫(P((uK,u∂K)) - uK)*vdΩ
\PiK =LocalProjector((m,l),UKR,VKR)

# Face term
m(ur,v) = ∫u*vd∂K
l( (uK,u∂K), v)=∫(P((uK,u∂K)) - u∂K)*vd∂K
\Pi∂K =LocalProjector((m,l),UKR,VKR)

# stab term in the global operator
a(u,v)=∫(∇(P(v))⋅∇(P(u)))dΩ + ∫\Pi∂K(u)*\Pi∂K(v)d∂K + ∫\PiK(u)*\PiK(v)dΩ
```

amartinhuertas committed:
First complete implementation of HHO gradient reconstruction operator

amartinhuertas committed:
Modularizing LocalFEOperatorTests.jl for better readability

amartinhuertas committed:
Added the necessary logic to support the reduction operator
(i.e., cell-wise + cell-boundary-wise L2 projection of bulk FE
functions)

amartinhuertas committed:
Potential reconstruction operator validated

I took a k+1-order polynomial from a H1-conforming
space, reduced to the k-order hybrid space, and
reconstructed it again to the k+1-th order non-conforming
space. The first and the last match.

amartinhuertas committed:
Difference operators working. All building blocks ready for HHO.

amartinhuertas committed:
Typo fix

amartinhuertas committed:
First complete implementation of HHO method for the Poisson problem.
TO-DEBUG: Currently it does not solve a problem with manufactured
solution in the solution FE space.

amartinhuertas committed:
Primal HHO for Poisson problem working!

amartinhuertas committed:
Cosmetic changes in the driver

amartinhuertas committed:
Adding @time to tests

amartinhuertas committed:
* Added convergence analysis to HHO driver
* Added stabilization

amartinhuertas committed:
Lowest order HHO convergence analysis

Jai-Tushar committed:
Project.toml and minor issues with PoissonHHOtests.jl

Jai-Tushar committed:
fixed minor issues with PoissonHHOTests.jl

amartinhuertas committed:
Fixes in LocalFEOperators + update to Julia 1.9

amartinhuertas committed:
checkpoint. Not working

amartinhuertas committed:
primal HHO + Poisson working for order=0, order=1 and order=3.

It does not work for order=2 nor order=4, though, :-(((. No idea why not at this
point.

Also, the code is currently quite slow, I do not know why yet, but clearly
needs to be investigated.

amartinhuertas committed:
Splitting of LocalFEOperator into ReconstructionFEOperator and
ProjectionFEOperator

amartinhuertas committed:
* Fixed a bug (a wrong sign) in the definition of the n(.,.) bilinear
  form associated to ProjectionFEOperator. As of now, the implementation
  seems to be working for any polynomial order with analytical solution
  living in the FE space
* Developed functions required to evaluate the residual of the HH0
  method.

amartinhuertas committed:
* Tidying up reconstruction operator definition in PoissonHHoTests.
* Adding tests with manufactured solutions in FE space up to order 3.

amartinhuertas committed:
Removing unused template parameters from function definition statement

amartinhuertas committed:
Minor changes to PoissonHHO with performance in mind

amartinhuertas committed:
Changes required to support latest version of Gridap (0.18)
